### PR TITLE
[ir/tests] Add allo.bmm() op and test cases

### PR DESCRIPTION
**Pull request by Zhichenzzz**
## Description ##
This PR adds the `build_bmm()` frontend feature to support the backend function `allo.bmm()`, fixes `no_init` initial values in `build_init()` and more diverse tests by using `linalg` dialect:

```python
@staticmethod
def build_Matmul(ctx, attr, new_args):
def build_init_zero(ctx, init_op, type):
```

### Changes and examples ###
- [Frontend] Support using the function of `allo.bmm()` like this. Besides, you can also nest this func:

``` python
def kernel(
    A: int32[M, N, K], B: int32[M, K, N], C: int32[M, N, K]
) -> int32[M, N, K]:
    D = allo.bmm(allo.bmm(A, B), C)
    return D
```
The generated IR displays:
``` mlir
module {
  func.func @kernel(%arg0: memref<10x25x20xi32>, %arg1: memref<10x20x25xi32>, %arg2: memref<10x25x20xi32>) -> memref<10x25x20xi32> {
    %0 = memref.alloc() : memref<10x25x25xi32>
    %c0_i32 = arith.constant 0 : i32
    linalg.fill ins(%c0_i32 : i32) outs(%0 : memref<10x25x25xi32>)
    linalg.batch_matmul ins(%arg0, %arg1 : memref<10x25x20xi32>, memref<10x20x25xi32>) outs(%0 : memref<10x25x25xi32>)
    %1 = memref.alloc() : memref<10x25x20xi32>
    %c0_i32_0 = arith.constant 0 : i32
    linalg.fill ins(%c0_i32_0 : i32) outs(%1 : memref<10x25x20xi32>)
    linalg.batch_matmul ins(%0, %arg2 : memref<10x25x25xi32>, memref<10x25x20xi32>) outs(%1 : memref<10x25x20xi32>)
    return %1 : memref<10x25x20xi32>
  }
}
```
After building:
``` mlir
module {
  func.func @kernel(%arg0: memref<10x25x20xi32>, %arg1: memref<10x20x25xi32>) -> memref<10x25x20xi32> {
    %c0_i32 = arith.constant 0 : i32
    %c1_i32 = arith.constant 1 : i32
    %alloc = memref.alloc() {name = "C"} : memref<10x25x20xi32>
    affine.for %arg2 = 0 to 10 {
      affine.for %arg3 = 0 to 25 {
        affine.for %arg4 = 0 to 20 {
          %0 = affine.load %arg0[%arg2, %arg3, %arg4] {from = "A"} : memref<10x25x20xi32>
          %1 = arith.addi %0, %c1_i32 : i32
          affine.store %1, %alloc[%arg2, %arg3, %arg4] {to = "C"} : memref<10x25x20xi32>
        } {loop_name = "k"}
      } {loop_name = "j"}
    } {loop_name = "i", op_name = "S_i_j_k_0"}
    %alloc_0 = memref.alloc() : memref<10x25x25xi32>
    affine.for %arg2 = 0 to 10 {
      affine.for %arg3 = 0 to 25 {
        affine.for %arg4 = 0 to 25 {
          affine.store %c0_i32, %alloc_0[%arg2, %arg3, %arg4] : memref<10x25x25xi32>
        }
      }
    }
    affine.for %arg2 = 0 to 10 {
      affine.for %arg3 = 0 to 25 {
        affine.for %arg4 = 0 to 25 {
          affine.for %arg5 = 0 to 20 {
            %0 = affine.load %arg0[%arg2, %arg3, %arg5] : memref<10x25x20xi32>
            %1 = affine.load %arg1[%arg2, %arg5, %arg4] : memref<10x20x25xi32>
            %2 = affine.load %alloc_0[%arg2, %arg3, %arg4] : memref<10x25x25xi32>
            %3 = arith.muli %0, %1 : i32
            %4 = arith.addi %2, %3 : i32
            affine.store %4, %alloc_0[%arg2, %arg3, %arg4] : memref<10x25x25xi32>
          }
        }
      }
    }
    %alloc_1 = memref.alloc() : memref<10x25x20xi32>
    affine.for %arg2 = 0 to 10 {
      affine.for %arg3 = 0 to 25 {
        affine.for %arg4 = 0 to 20 {
          affine.store %c0_i32, %alloc_1[%arg2, %arg3, %arg4] : memref<10x25x20xi32>
        }
      }
    }
    affine.for %arg2 = 0 to 10 {
      affine.for %arg3 = 0 to 25 {
        affine.for %arg4 = 0 to 20 {
          affine.for %arg5 = 0 to 25 {
            %0 = affine.load %alloc_0[%arg2, %arg3, %arg5] : memref<10x25x25xi32>
            %1 = affine.load %alloc[%arg2, %arg5, %arg4] : memref<10x25x20xi32>
            %2 = affine.load %alloc_1[%arg2, %arg3, %arg4] : memref<10x25x20xi32>
            %3 = arith.muli %0, %1 : i32
            %4 = arith.addi %2, %3 : i32
            affine.store %4, %alloc_1[%arg2, %arg3, %arg4] : memref<10x25x20xi32>
          }
        }
      }
    }
    return %alloc_1 : memref<10x25x20xi32>
  }
}
```
- [Frontend fix] Add `build_init_zero()` function in ir/builder.py, which uses linalg dialect to set values to zeros.

- [Test cases] Make a new file `test_linalg.py` for linalg dialect building. Moreover, add new complex tests to ensure the robustness of codes.
## Checklist ##

- [x] PR's title starts with a category (e.g. [Bugfix], [IR], [Builder], etc)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage (It would be better to provide ~2 different test cases to test the robustness of your code)
- [x] Code is well-documented

### CODE REVIEW

File: allo/ir/builder.py [hunk 1]
```diff
@@ -941,39 +943,61 @@ def build_Call(ctx, node):
...
+        # matrix shape
         dtype = ShapedType(new_args[0].type).element_type
         argAshape = ShapedType(new_args[0].type).shape
         argBshape = ShapedType(new_args[1].type).shape
-        if len(argAshape) != 2 or len(argBshape) != 2:
-            raise RuntimeError("Only support two 2D matrix multiplication")
-        shape = (argAshape[0], argBshape[1])
+        if attr == "matmul":
+            if len(argAshape) != 2 or len(argBshape) != 2:
+                raise RuntimeError("Only support two 2D matrix multiplication")
```
Comment (chhzh123) [inline allo/ir/builder.py]: "Only support matrix multiplication of two 2D inputs"

File: allo/ir/builder.py [hunk 2]
```diff
@@ -941,39 +943,61 @@ def build_Call(ctx, node):
...
+            shape = (argAshape[0], argBshape[1])
+        if attr == "bmm":
+            if len(argAshape) != 3 or len(argBshape) != 3:
+                raise RuntimeError("Only support two 3D batch matrix multiplication")
```
Comment (chhzh123) [inline allo/ir/builder.py]: "Only support batched matrix multiplication of two 3D inputs"

File: allo/ir/builder.py [hunk 3]
```diff
@@ -941,39 +943,61 @@ def build_Call(ctx, node):
...
+            if attr == "bmm":
+                linalg_d.batch_matmul(
+                    new_args[0],
+                    new_args[1],
+                    outs=[alloc_op],
+                )
+        return alloc_op
+
+    @staticmethod
+    def build_init(ctx, init_op, dtype):
```
Comment (chhzh123) [inline allo/ir/builder.py]: `build_init_zero`

File: tests/test_linalg.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+                0.5
+                * A1[i, j]
+                * (
+                    1.0
+                    + allo.tanh(
+                        allo.sqrt(2.0 / 3.1415926)
+                        * (A1[i, j] + 0.044715 * allo.power(A1[i, j], 3.0))
+                    )
+                )
+            ) - D[i, j]
```
Comment (chhzh123) [inline tests/test_linalg.py]: Is it a GeLU function? Can you add a line to comment on that?

File: allo/ir/builder.py [hunk 5]
```diff
@@ -623,6 +623,8 @@ def build_AnnAssign(ctx, node):
                         with ip:
                             # pylint: disable=unexpected-keyword-arg
                             linalg_d.fill(rhs.result, outs=[alloc_op.result])
+                    else:
+                        ASTTransformer.build_init(ctx, alloc_op, ele_type)
```
Comment (chhzh123) [inline allo/ir/builder.py]: Why did you add this line? If a variable is not initialized with a value, we should keep it uninitialized.

File: tests/test_parser.py [hunk 6]
```diff
@@ -749,7 +749,7 @@ def test_const_tensor_int():
     def kernel() -> int32[2, 2]:
         cp1: int32[2, 2] = [[1, 2], [3, 4]]
         cp2: int32[2, 2] = [[1, 2], [3, 4]]
-        res: int32[2, 2] = 0
+        res: int32[2, 2]
```
Comment (chhzh123) [inline tests/test_parser.py]: Why did you remove the initial value? This program will become incorrect since you change the behavior. Notice if a variable is not given an RHS value, we should keep it uninitialized, which is important for specific hardware design -- we don't need to initialize every buffer in the very beginning.

File: tests/test_parser.py [hunk 7]
```diff
@@ -767,9 +767,9 @@ def test_const_tensor_float():
     def kernel() -> float32[2, 3]:
         cp1: float32[2, 3] = [[1.05, 2.0, 3.5], [3.0, 4.0, 4.5]]
         cp2: float32[2, 3] = [[1.05, 2.0, 3.5], [3.0, 4.0, 4.5]]
-        res: float32[2, 3] = 0.0
+        res: float32[2, 3]
```
Comment (chhzh123) [inline tests/test_parser.py]: Same issue. Should provide an initial value.

File: tests/test_parser.py [hunk 8]
```diff
@@ -790,7 +790,7 @@ def test_const_tensor_int_vars():
     def kernel() -> int32[M, N]:
         cp1: int32[M, N] = np_0
         cp2: int32[M, N] = np_1
-        res: int32[M, N] = 1
+        res: int32[M, N]
```
Comment (chhzh123) [inline tests/test_parser.py]: Same

File: tests/test_parser.py [hunk 9]
```diff
@@ -878,6 +878,64 @@ def kernel() -> float32[M, K]:
...
+    s = allo.customize(kernel)
+    print(s.module)
+    f = s.build()
+    outs = np.zeros((M, N, K), dtype="int32")
+    outs = f(A, B, A)
+    out_1 = np.einsum("ijk,ikn->ijn", A, B)
+    out_2 = np.einsum("ijk,ikn->ijn", out_1, A)
+    np.testing.assert_allclose(outs, out_2, atol=1e-4)
+
+
```
Comment (chhzh123) [inline tests/test_parser.py]: Can you also remove these tests? As you have already moved them to the `test_linalg` file.

File: tests/test_parser.py (hunk 6 above)
Comment (Zhichenzzz) [reply tests/test_parser.py]: Oh sorry, I mistook the goal of initial value. I will recover its original expression.

File: tests/test_linalg.py (hunk 4 above)
Comment (Zhichenzzz) [reply tests/test_linalg.py]: Ok, there are some problems still.
